src/Package/Impl/ProjectSystem/RDebugLaunchProvider.cs

```diff
@@ -85,31 +87,126 @@ public RDebugLaunchProvider(ConfiguredProject configuredProject, IRInteractiveWo
+
+            var projectDir = Path.GetDirectoryName(project.FullName);
+            var projectName = Path.GetFileNameWithoutExtension(project.FullName);
+            var filter = new TransferFileFilter(filterString);
+
+            await Callbacks.WriteConsoleEx($"Remote destination: {remotePath}\n", OutputType.Output, CancellationToken.None);
+            await Callbacks.WriteConsoleEx($"File filter applied: {filterString}\n", OutputType.Output, CancellationToken.None);
+            await Callbacks.WriteConsoleEx("Compressing project files for transfer:\n", OutputType.Output, CancellationToken.None);
+
+            var compressedFilePath = await Task.Run(() => FileSystem.CompressDirectory(projectDir, (p) => {
+                Callbacks.WriteConsoleEx($"Compressing: {p}\n", OutputType.Output, CancellationToken.None).Wait();
+                return filter.Match(p);
+            }));
+
+            using (var fts = new FileTransferSession(Session, FileSystem)) {
+                await Callbacks.WriteConsoleEx("Transferring project to remote host...", OutputType.Output, CancellationToken.None);
+                var remoteFile = await fts.SendFileAsync(compressedFilePath);
+                await Session.EvaluateAsync<string>(Invariant($"rtvs:::save_project({remoteFile.Id}, {projectName.ToRStringLiteral()}, '{remotePath.ToRPath()}')"), REvaluationKind.Normal);
+            }
+
+            await Callbacks.WriteConsoleEx(" Completed.\n", OutputType.Output, CancellationToken.None);
+        }
+
+        private async Task<string> GetStartupFileAsync(bool transferFiles, EnvDTE.Project project) {
+            if (transferFiles && Session.IsRemoteSession) { // remote
+                var projectName = Path.GetFileNameWithoutExtension(project.FullName);
+                var remotePath = (await _properties.GetRemoteProjectPathAsync()).ToRPath();
+                var startUpFile = (await _properties.GetStartupFileAsync()).ToRPath();
+                return remotePath + projectName + "/" + startUpFile;
+            } else { // local
+                var projDir = Path.GetDirectoryName(project.FullName);
+                var startUpFile = await _properties.GetStartupFileAsync();
+                return Path.Combine(projDir, startUpFile);
+            }
+        }
+
+        private string GetRemoteSettingsFile(string localSettingsPath, string localProjectPath, string remoteProjectPath) {
+            return remoteProjectPath + localSettingsPath.Remove(0, localSettingsPath.Length);
+        }
+
+        private class TransferFileFilter {
+            List<string> filterPatterns;
+            public TransferFileFilter(string filterString) {
+                string[] patterns = filterString.Split(separator, StringSplitOptions.RemoveEmptyEntries);
+                filterPatterns = new List<string>(patterns);
+            }
+            private static string[] separator = { ";"};
+
+            public bool Match(string path) {
```

Comment:
Have a look at this:

https://www.nuget.org/packages/Microsoft.Extensions.FileSystemGlobbing

https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/Extensions/FileSystemGlobbing/Matcher/index.html#matcher-class